### ISSUE TITLE
Versioned Caching

### ISSUE THREAD

**Issue by mmccarty** [intake/intake#351]
Intake caching was added to avoid the need to re-download remote data each time a catalog is loaded.  Another use case for caching could be to provide versioning datasets for a given time.  This caching strategy would save a snapshot of the previous version each time the cache is refreshed.  The refresh would need to be triggered manually since Intake has no way of knowing when new data is available.  The user could specify a tag in the form of a string to be used later to restore that version of the data.  Alternatively, a version should be retrieved for a given timestamp.  Perhaps `up` and `down` methods for stepping forward and backward through time.  This would be quite useful in my work, but I'm wondering if others agree?

**Comment by martindurant** [comment]
Related to #337 , which suggests the reverse , in which versioning might be provided by an external service, and Intake's job is to reflect the versioning information, and so have an automatic way to know whether the currently stored data is the most recent or not. Forms of this simpler than DVC might be:
- versioned S3 buckets
- any git (or other VCS) repo

Other ways in which "currentness" might be checked or tags/labels established: file-systems which expose a timestamp (like NFS, SSH); filesystems which expose a checksum (like S3, GCS, Azure).

Not to muddy the water, but also worth a thought: some services like SQL tables might guarantee incremental updates and some monotonically increasing index. Then we need to know the max index now and last time we looked and can just get the diff. Maybe this point is relevant to persisting rather than caching.

**Comment by mmccarty** [comment]
These forms seem more like versioned data sources than a caching mechanism.  The key difference here is that Intake would allow the user specify versions.  

(Have we thought about git (or other VCS) repos as data sources?)

**Comment by martindurant** [comment]
Git is certainly a caching mechanism, and I would like to see it added; it would be similar to DAT, you have a reference, and therein some files. The difference between that versioning and the one you are talking about, is whether we are in charge of locating and labelling the versions. In either case, you can list available versions, pick the one you want, and mark particular states as useful.

**Comment by mmccarty** [comment]
Agreed.  I'm thinking Git might be useful as a first pass on this.  I believe we would be able to support many of the existing caching strategies.

**Comment by michaelaye** [comment]
hm, doesn't the fact that conda enforces versioning already solve the problem on that side? But I guess you don't want to go down the "enforcing conda" route?

**Comment by martindurant** [comment]
Yes and no...
If you have data files directly in the package, or the package contains a remote URL which points to something that is itself versioned (e.g., using versioned buckets on S3), then yes, that would be enough.

However, it also common that you are loading from some remote file that can change whenever, even if the *spec* for what and how to read is in a conda package. If you would like to keep versions of that data at various time-points, you need something additional.

**Comment by michaelaye** [comment]
I see, true. 
Related to this then, I think `intake` should determine by nature of the source if it's really versioned and set a flag called `versioned` (or `safe_versioned` ?)  to True or False, depending on the outcome. It will inform the user of the nature of versioning of the provided data-set.

**Comment by michaelaye** [comment]
Incidentally, our institute just does a cloud services review, and if S3 buckets could be used for safe and versioned public data catalog dissemination, I'm sure that would be a great incentive for the institute to offer these.

**Comment by martindurant** [comment]
I imagine it's something done at the caching and/or persist layer: if it's files, we may or may not keep versions in the local cache; if it's an external service like git or DAT, we surface any versions it knows about. 

The "intermediate" situation of external file versions like S3 is more complex, since each file has a different set of versions; in this case, I would think there needs to be a separate route for getting the available versions and passing the appropriate parameter to the file loader.

**Comment by michaelaye** [comment]
does intake provide a way to get to the conda package version?

**Comment by michaelaye** [comment]
or in general any version number of the data catalog?

**Comment by martindurant** [comment]
Intake does not particularly know where the catalog files it loads comes from, whether installed by a conda package or not. However, you can add arbitrary information into the metadata section of a catalog file:
```
name: name_in_cat
metadata:
  test: true
  version: 2.5
```
In your case, you would put this information into the *inner* cat file, and you could even write your conda install script to insert the line based on the environment variable version at the time of installation.

**Comment by michaelaye** [comment]
Ok, so I figured out from here:
https://docs.conda.io/projects/conda-build/en/latest/user-guide/environment-variables.html
that I have `PKG_VERSION` available as a variable at the time of installation. Could you give me a hint how to write that into the catalog files within the build.sh script? Should I create a Python script reading, changing and writing out the catalog JSON that is being called by build.sh?

**Comment by martindurant** [comment]
It could be quite easily done with `sed` on posix http://www.theunixschool.com/2012/06/insert-line-before-or-after-pattern.html

but maybe we should stick to python
```
import os
import sys

def add_conda_version(infile):
    verline = '  conda_package_version: %s\n' % os.environ['PKG_VERSION']
    text = open(infile).read().replace(
        'metadata:\n', 'metadata:\n' + verline, 1)
    open(infile, 'w').write(text)


if __name__ == '__main__':
    add_conda_version(sys.argv[1])
```

Something more thorough could be made and stored in the Intake repo as a helper.
